File: src/util/entities/CloudAttachment.ts
/*
	Spacebar: A FOSS re-implementation and extension of the Discord.com backend.
	Copyright (C) 2023 Spacebar and Spacebar Contributors

	This program is free software: you can redistribute it and/or modify
	it under the terms of the GNU Affero General Public License as published
	by the Free Software Foundation, either version 3 of the License, or
	(at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU Affero General Public License for more details.

	You should have received a copy of the GNU Affero General Public License
	along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

import { Column, Entity, JoinColumn, ManyToOne, RelationId } from "typeorm";
import { BaseClass } from "./BaseClass";
import { dbEngine } from "../util/Database";
import { User } from "./User";
import { Channel } from "./Channel";

@Entity({
	name: "cloud_attachments",
	engine: dbEngine,
})
export class CloudAttachment extends BaseClass {
	// Internal tracking metadata
	@Column({ name: "user_id", nullable: true })
	@RelationId((att: CloudAttachment) => att.user)
	userId: string;

	@JoinColumn({ name: "user_id" })
	@ManyToOne(() => User, { nullable: true, onDelete: "SET NULL" })
	user?: User;

	@Column({ name: "channel_id", nullable: true })
	@RelationId((att: CloudAttachment) => att.channel)
	channelId?: string; // channel the file is uploaded to

	@JoinColumn({ name: "channel_id" })
	@ManyToOne(() => Channel, { nullable: true, onDelete: "SET NULL" })
	channel?: Channel; // channel the file is uploaded to

	@Column({ name: "upload_filename" })
	uploadFilename: string;

	// User-provided info
	@Column({ name: "user_attachment_id", nullable: true })
	userAttachmentId?: string;

	@Column({ name: "user_filename" })
	userFilename: string; // name of file attached

	@Column({ name: "user_file_size", nullable: true })
	userFileSize?: number; // size of file in bytes

	@Column({ name: "user_original_content_type", nullable: true })
	userOriginalContentType?: string;

	@Column({ name: "user_is_clip", nullable: true })
	userIsClip?: boolean; // whether the file is a clip

	// Actual file info, initialised after upload
	@Column({ nullable: true })
	size?: number; // size of file in bytes

	@Column({ nullable: true })
	height?: number; // height of file (if image)

	@Column({ nullable: true })
	width?: number; // width of file (if image)

	@Column({ name: "content_type", nullable: true })
	contentType?: string;

	// @BeforeRemove()
	// onDelete() {
	// 	return deleteFile(new URL(this.url).pathname);
	// }
}

File: extra/admin-api/Models/Spacebar.Models.Db/Models/CloudAttachment.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using Microsoft.EntityFrameworkCore;

namespace Spacebar.Models.Db.Models;

[Table("cloud_attachments")]
public partial class CloudAttachment
{
    [Key]
    [Column("id", TypeName = "character varying")]
    public string Id { get; set; } = null!;

    [Column("user_id", TypeName = "character varying")]
    public string? UserId { get; set; }

    [Column("channel_id", TypeName = "character varying")]
    public string? ChannelId { get; set; }

    [Column("upload_filename", TypeName = "character varying")]
    public string UploadFilename { get; set; } = null!;

    [Column("user_attachment_id", TypeName = "character varying")]
    public string? UserAttachmentId { get; set; }

    [Column("user_filename", TypeName = "character varying")]
    public string UserFilename { get; set; } = null!;

    [Column("user_file_size")]
    public int? UserFileSize { get; set; }

    [Column("user_original_content_type", TypeName = "character varying")]
    public string? UserOriginalContentType { get; set; }

    [Column("user_is_clip")]
    public bool? UserIsClip { get; set; }

    [Column("size")]
    public int? Size { get; set; }

    [Column("height")]
    public int? Height { get; set; }

    [Column("width")]
    public int? Width { get; set; }

    [Column("content_type", TypeName = "character varying")]
    public string? ContentType { get; set; }

    [ForeignKey("ChannelId")]
    [InverseProperty("CloudAttachments")]
    public virtual Channel? Channel { get; set; }

    [ForeignKey("UserId")]
    [InverseProperty("CloudAttachments")]
    public virtual User? User { get; set; }
}
